docs/pallas/tpu/core_map.md

```diff
@@ -0,0 +1,357 @@
+
+ * Start your top-level JAX code with HBM refs, and allocate output refs if needed.
+
+ * Use `pl.core_map`, which takes the TensorCore mesh, to start per-core programming.
+
+    * You will need `collective_id` for the barrier semaphore.
+
+ * Inside `pl.core_map`, invoke `pl.run_scoped` to allocate per-core scratch spaces (VMEM and semaphores) and run the local kernel.
+
+
+```python
+input_shape = (32, 256)
+local_vmem_shape = (32 // num_devices, 256 // num_cores)
+in_spec = jax.P('device', None)
+sharding = NamedSharding(mesh, in_spec)
+
+@jax.jit
+@partial(jax.shard_map, mesh=mesh, in_specs=in_spec, out_specs=in_spec,
+         check_vma=False)
+def swap_cores(x):
+  # Get buffers out of the input and output
+  x_hbm_ref, o_hbm_ref = jax.tree.map(jax.new_ref, (x, jax.lax.empty(x.shape, x.dtype)))
+
+  @pl.core_map(tc_mesh, compiler_params=pltpu.CompilerParams(collective_id=0))
+  def _():
+    pl.run_scoped(
+        partial(swap_cores_kernel, x_hbm_ref, o_hbm_ref),
+        *([pltpu.VMEM(local_vmem_shape, x.dtype)] * 3),  # VMEM allocations
+        *([pltpu.SemaphoreType.DMA] * 3),                # semaphores
+    )
+  return o_hbm_ref[...]
+
+
+x = jax.random.normal(jax.random.key(0), input_shape, jnp.float32)
+x = jax.device_put(x, sharding)
+y = swap_cores(x)
+
+np.testing.assert_array_equal(y[:, 128:], x[:, :128] * 2)
+np.testing.assert_array_equal(y[:, :128], x[:, 128:] * 2)
+```
+
+### Save the boilerplate
+
+For `jax>=0.8.1`, you can use decorator `pl.kernel` to wrap boilerplates such as `core_map`, `run_scoped`, and output buffer allocation.
+
+Note that this should run inside any `jax.shard_map` you may have at the top level.
+
+
+```python
+@jax.jit
+@partial(jax.shard_map, mesh=mesh, in_specs=in_spec, out_specs=in_spec, check_vma=False)
+def swap_cores(x):
+  scratch_shapes = [pltpu.VMEM(local_vmem_shape, x.dtype)] * 3 + [pltpu.SemaphoreType.DMA] * 3
+  return pl.kernel(swap_cores_kernel, out_shape=x, mesh=tc_mesh,
+                   scratch_shapes=scratch_shapes,
+                   compiler_params=pltpu.CompilerParams(collective_id=0))(x)
+
+y = swap_cores(x)
+np.testing.assert_array_equal(y[:, 128:], x[:, :128] * 2)
+np.testing.assert_array_equal(y[:, :128], x[:, 128:] * 2)
+```
+
+## Pipelining with `core_map`
+
+Note that the kernel above only does simple copies and compute, without automatic pipelining via Pallas `grid` and `BlockSpec`. To do pipelining inside `core_map`, use `pltpu.emit_pipeline` inside the core-local kernel.
+
+**Automatically parallelize work amongst cores**
+
+The simple way is to annotate a block axis as `pltpu.PARALLEL`, and Pallas will automatically parallelize work along this axis. Both `pl.pallas_call` and `pltpu.emit_pipeline` supports this, via arguments `core_axis` and `dimension_semantics`. The `pallas_call` example was [in another guide](https://docs.jax.dev/en/latest/pallas/tpu/pipelining.html#tpus-in-megacore-configuration), and the `emit_pipeline` case is shown below.
```

Comment:
Add a little paragraph:

When the PARALLEL annotation is provided and the corresponding grid dimension will be logically split and executed on separate cores. (The exact semantics of which grid dimensions are executed on which core is guaranteed).